Data related features

lexize opened the pull request:
This pull requests adds some data related features like FileAPI. Full list of changes:
### File API
API made to provide features to work with files.
Operations related to files can be done only in `figura/data` folder an only on host.
### Data API
This API adds features related to advanced data manipulation, as input/output streams, data readers for input streams, data providers for output streams, and also byte buffers.
### Json API
API that adds some types and functions for JSON serialization and deserialization.
### Resources API
Allows you to store raw file data from your avatar folder in avatar itself. To include resource in your avatar you must add resource path/[Glob  pattern](https://docs.oracle.com/javase/tutorial/essential/io/fileOps.html#glob) to `resources` array in your metadata
### Networking API
Everything related to networking. Right now it is only HTTP requests, TCP is planned.
To use Networking API you will need to enable it in settings and in avatar permissions, and also setup your Network Filter, by setting restriction mode to "Whitelist", "Blacklist" or "None" and filling your filter with links and matching rules.
Networking is being logged to `figura/logs`, you can toggle logging off in settings.

### Existing code changes:
#### Config:
Added an interface named SerializableConfig, that provides methods for direct serialization and deserialization from JsonElements
#### Lua Runtime
Changed `pairs` and `ipairs` functions to make them able to be work with userdata, and not only tables
#### ConfigAPI
ConfigAPI now uses `figura/config` folder instead of `figura/data`. `figura/data` is now used by FileAPI

lexize committed:
Data update :woozy_face:

* Added a new interface to ConfigType: SerializableConfig
* Added a new config options: Enable networking, Networking restriction, Networking filter (WIP)
* Added a new permission: Allow networking
* New API's: NetworkingAPI (lacks documentation right now), HttpAPI, DataAPI (very WIP)
* New types: Future, InputStream, OutputStream, Provider, Reader

lexize committed:
Docs momento

* Added new APIs to a docs command.
* Added readers and providers constant to DataAPI
* Added StringReaders and StringProviders for all standard java encodings.
* Added translation strings for classes

lexize committed:
Files

* Added FileAPI
* New avatar metadata option: dataFolder
* Added translation to all fork APIs and classes

lexize committed:
No networking :pensive:

Removed everything related to networking.

Added two classes: StreamReader and StreamWriter (this one is still empty)

Added new method to DataAPI: createBufferInputStream - creates an input stream from provided data buffer.

lexize committed:
Json thingies

Started writing json things

lexize committed:
Json thingis part two

Implemented most of JSON functions that was planned
Added pairs and ipairs for FiguraJsonArray and FiguraJsonObject
Moved everything related to json from `data` to separate global `json`
Changed a method in LuaUtils that returns lua value as a json element, so it can process FiguraJsonArray and FiguraJsonObjects
Renamed StreamReader and StreamWriter to ReadUtils and WriteUtils

lexize committed:
Me forgora :skull:

Fixed endianess in ReadUtils, implemented things in WriteUtils

lexize committed:
Who tf is ReadUtils and WriteUtils?

Created new class: FiguraBuffer

Removed ReadUtils and WriteUtils and moved methods from them into FiguraBuffer

lexize committed:
Docs momento part two

(probably) Finished everything related to docs.
Added readString and writeString to FiguraBuffer
Added LuaNotNil where it is needed,
Added toString methods to all added types so it just prints type's name

lexize commented:
:blobfoxfloofcofe:

Covkie commented:
> should be merged when backend v3 will be too

-skylar

lexize committed:
More methods in InputStream and OutputStream

Created interfaces FiguraReadable and FiguraWritable and moved most of logic of FiguraBuffer to them
Made FiguraInputStream and FiguraOutputStream implement FiguraReadable and FiguraWritable
Removed old translations as i will fucking die if i try to reformat everything :skull: (i backed everything up)

lexize commented:
Im sorry for force pushes but my git (or maybe this is github problem) refuses to push commit because blah blah your branch is older than remote branch, while it is not. Im so tired of this so im just force pushing shit

Covkie commented:
I don't know if this is an issue with me not knowing how to navigate the mobile UI but it seems the en_us.json entries for this pr were lost with the force push?

lexize commented:
No i just removed the en_us changes from pr because if i would try to make them match current en_us formatting I would die

Covkie committed:
Restore en_us.json entries

restores lang file entries to be compatible with the new format used in en_us and all other lang files

please let me know if I missed anything

Covkie commented on common/src/main/resources/assets/figura/lang/en_us.json:
```diff
@@ -617,6 +620,7 @@
     "figura.docs.globals.pings": "The global instance of PingAPI",
     "figura.docs.globals.textures": "The global instance of the TextureAPI",
     "figura.docs.globals.config": "The global instance of the ConfigAPI",
+    "figura.docs.globals.net": "The global instance of the NetworkingAPI",
```
This does not have any use and I assume you forgot to remove it in commit `30ef162f`(No networking 😔). I included it for posterity's sake.

lexize replied:
Wait why tf is it here? :skull: 
As i remember i literally made my pr use en_us from main branch

Edit: Yeah this en_us is from old commit, it shouldnt be here at all.

Covkie replied:
I took the latest commit still with the en_us translations that you made and modified the entries to be compatible with the current format so the doc entries and everything else are translatable.

lexize replied:
Ah, ok, thank you. Yes this thing is not used anymore as i removed everything related to Networking, let me remove it rq

lexize committed:
Fixed shit, added some other shit

Removed "globals.net" from translation as it is not used
Added "globals.file", "globals.data", and "globals.json"

weblate committed:
Translations update from Hosted Weblate (#98)

lexize committed:
Resources API, full rewrite of translations

\## Added new API: ResourcesAPI
Files in avatar folder can be included as avatar resources if you add them in `resources` field in avatar metadata. Paths are relative to avatar folder. Glob file matching is supported Example:
```json
{
	...
	"resources": [
		"your_file_name.ext",
		"your_folder/*"
	]
	...
}
```

Fully rewrote translation strings for everything in PR

lexize committed:
Buffer close

Added "close" and "isClosed" functions to FiguraBuffer
Added "Buffers Count" permission

lexize committed:
Fixed some things

lexize committed:
Yes networking 🥳

Reverted changes of commit 85f9b26d1f35d78533c33f7040b6264902812809
Started implementing Network Filter screen

lexize committed:
FileAPI fix
Added argument docs for FileAPI
Added LuaWhitelist and LuaTypeDoc annotations to FiguraFuture

lexize committed:
FileAPI fix №2

lexize committed:
Optional reader and writer

Made readers and writers in FileAPI to be optional

lexize committed:
No readers and providers

Removed all the usages of readers and providers

lexize committed:
No data folder requirement

Removed dataFolder field from avatar metadata

lexize committed:
Changed Config root folder
Chaged ConfigAPI root folder from figura/data to figura/config

lexize committed:
HttpRequestsAPI fix
Added subclasses of HttpRequestsAPI to whitelisted classes

lexize committed:
Networking logging

Added networking logging and settings related to logging
Added another network filter mode - contains

lexize committed:
Merge remote-tracking branch 'base/1.20' into 1.20

lexize committed:
Minor logging fixes

lexize committed:
Networking docs

Added more docs

lexize committed:
Buffer fixes

Fixed some things in interface FiguraReadable and in a Buffer itself

lexize committed:
Another buffer fix

lexize committed:
Another buffer fixes (i hope last ones)

lexize committed:
FiguraSocket

Added new methods to FiguraBuffer, FiguraInputStream: readBase64 and readByteArray
Added new methods to FiguraBuffer, FiguraOutputStream: writeBase64, writeByteArray

Added FiguraSocket and SocketsAPI

Added toString to FiguraFuture

lexize committed:
My bad

lexize committed:
Close buffers on avatar unload

lexize committed:
Some fixes

Fixed scrolling issue
Fixed wrong filter matchers IDs
Fixed crash on avatar clear related to closing buffers and sockets
Fixed LinkNotAllowedException
Fixed resources path, now it always uses "/" as separator
Now you can access resource stream with resources["path/to/resource"]

lexize committed:
Json fix

I AM SO FUCKING STUPID IM SORRY

lexize committed:
Async and Json Fixes

Added hasError, throwError, and getOrError to FiguraFuture. isDone returns true both in case if Future was executed successfully or with error.
Made net.socket:open return Future instead to prevent game freezes with slow internet
Added JsonAPI to json global docs
Fixed translation string in JsonAPI

lexize committed:
Removed methods from FiguraInputStream and FiguraOutputStream

Commit still has old bytes-to-string conversion method

lexize committed:
Me forgora

lexize committed:
Forgora fix

lexize committed:
Buffer fix
Idk how i've missed this :skull:

lexize committed:
Buffer fix #2

I really have no idea why does it crash in some cases

lexize committed:
Buffer fix

lexize committed:
Asyncing unasyncable

Added two methods in FiguraInputStream: readAsync and isAsyncOnly

lexize committed:
Fuck readables and writables

Removed FiguraReadable and FiguraWritable as they are not needed anymore

lexize committed:
Merge branch '1.20' of github.com:FiguraMC/Figura into 1.20

Covkie committed:
Merge branch '1.20' into 1.20
